File: app/lib/replayTiming.ts
// Play one storm-hour in 30 seconds (120x), with an 8-second floor for short
// storms and a 90-second ceiling so a full day's activity stays easy to watch.
export const REPLAY_MS_MIN = 8_000;
export const REPLAY_MS_MAX = 90_000;
export const REPLAY_MS_PER_STORM_MIN = 500;

export function computeReplayDurationMs(spanMs: number): number {
  if (!Number.isFinite(spanMs) || spanMs <= 0) return REPLAY_MS_MIN;
  return Math.min(
    REPLAY_MS_MAX,
    Math.max(REPLAY_MS_MIN, (spanMs / 60_000) * REPLAY_MS_PER_STORM_MIN),
  );
}

// During playback a strike counts as "fresh" for ~1.2 real seconds, scaled by
// how compressed real time is relative to replay time.
export function computeFreshMs(spanMs: number, replayMs: number): number {
  if (!Number.isFinite(spanMs) || !Number.isFinite(replayMs) || spanMs <= 0 || replayMs <= 0) return 0;
  return (spanMs / replayMs) * 1200;
}

export function cutoffForProgress(progress: number, minTime: number, maxTime: number): number {
  const p = Math.min(1, Math.max(0, progress));
  return minTime + p * (maxTime - minTime);
}

export function progressForCutoff(cutoff: number, minTime: number, maxTime: number): number {
  const span = maxTime - minTime;
  if (span <= 0) return 0;
  return Math.min(1, Math.max(0, (cutoff - minTime) / span));
}

File: lightning-stats/lib/utils/replay_timing.dart
// Ported 1:1 from app/lib/replayTiming.ts.
// Playback lasts ~2s per storm-minute, clamped so short storms stay watchable
// and multi-hour storms don't drag on forever.
const int kReplayMsMin = 10000;
const int kReplayMsMax = 300000;
const int kReplayMsPerStormMin = 2000;

int computeReplayDurationMs(int spanMs) {
  final scaled = (spanMs / 60000) * kReplayMsPerStormMin;
  return scaled.clamp(kReplayMsMin, kReplayMsMax).round();
}

/// During playback a strike counts as "fresh" for ~1.2 real seconds, scaled by
/// how compressed real time is relative to replay time.
double computeFreshMs(int spanMs, int replayMs) => (spanMs / replayMs) * 1200;

int cutoffForProgress(double progress, int minTime, int maxTime) {
  final p = progress.clamp(0.0, 1.0);
  return (minTime + p * (maxTime - minTime)).round();
}

double progressForCutoff(int cutoff, int minTime, int maxTime) {
  final span = maxTime - minTime;
  if (span <= 0) return 0;
  return ((cutoff - minTime) / span).clamp(0.0, 1.0);
}
